### PR TITLE
fix(pyclient ci): validate version arg in create-release

### PR DESCRIPTION
**Pull request by dshemetov**
### Summary:

Fix the release CI to accept any version number and major, minor, patch as its argument. My test script for the versionNumber logic is below.

Notes:

- [set-output was deprecated](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/), so moved over to the new thing.
- tried [bump-my-version](https://github.com/callowayproject/bump-my-version), but it wasn't a significant improvement and had dependency resolver issues in `pip install -r requirements.api.txt -r requirements.dev.txt` in our Dockerfile.
- [regex tests](https://regexr.com/81dmt). 

```sh
# Try (major|minor|patch|5.3.0|blah|3.2.2.2)
versionName=major
allowed_pattern="^(major|minor|patch|[0-9]+\.[0-9]+\.[0-9]+)$"

if [[ ! $versionName =~ $allowed_pattern ]]; then
  echo "\nInvalid version name: $versionName"
  exit 1
fi

if [[ $versionName =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
  NEXT_TAG="$versionName"
else
  NEXT_TAG=$(bump2version --dry-run --list $versionName | grep ^new_version | sed -r s,"^.*=",,)
fi
echo "next_tag=$NEXT_TAG"
```

### Prerequisites:

- [x] Unless it is a documentation hotfix it should be merged against the `dev` branch
- [x] Branch is up-to-date with the branch to be merged with, i.e. `dev`
- [x] Build is successful
- [x] Code is cleaned up and formatted

### CODE REVIEW

File: .github/workflows/create-release.yml [hunk 1]
```diff
@@ -27,10 +27,23 @@ jobs:
...
+          if [[ ! ${{ github.event.inputs.versionName }} =~ $allowed_pattern ]]; then
+            echo "\nInvalid version name: ${{ github.event.inputs.versionName }}"
+            exit 1
+          fi
+
+          if [[ ${{ github.event.inputs.versionName }} =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+            echo "${{ github.event.inputs.versionName }}"
+          else
+            echo $(bump-my-version show-bump | grep ${{ github.event.inputs.versionName }} | sed -r s,"^.*─ ",,)
+          fi
```
Comment (melange396) [inline .github/workflows/create-release.yml:46]: `bump-my-version` doesnt look a whole lot better than `bump2version`...  what if we just keep using bump2 but with the logic to either extract a numeric version from the input w/ your regex, or otherwise try to generate a new numeric version?

```suggestion
          python -m pip install bump2version
          if [[ ${{ github.event.inputs.versionName }} =~ "^[0-9]+\.[0-9]+\.[0-9]+$" ]]; then
            NEXT_TAG="${{ github.event.inputs.versionName }}"
          else
            NEXT_TAG=$(bump2version --dry-run --list ${{ github.event.inputs.versionName }} | grep ^new_version | sed -r s,"^.*=",,)
          fi
          bump2version --new-version $NEXT_TAG _ignored_arg_
          echo "next_tag=$NEXT_TAG" >> $GITHUB_OUTPUT
```

and then below:
`bump2version --allow-dirty --new-version ${{ steps.version.outputs.next_tag }} _ignored_arg_`

Comment (dshemetov) [reply .github/workflows/create-release.yml]: Sure, I pushed a fix.

**Review by rzats — APPROVED** [review]
lgtm!

File: src/client/packaging/pypi/pyproject.toml [hunk 2]
```diff
@@ -42,3 +42,27 @@ dependencies = ["aiohttp", "delphi-utils", "requests>=2.7.0", "tenacity"]
 [project.urls]
 "Homepage" = "https://github.com/cmu-delphi/delphi-epidata"
 "Changelog" = "https://github.com/cmu-delphi/delphi-epidata/blob/main/src/client/packaging/pypi/CHANGELOG.md"
+
```
Comment (melange396) [inline src/client/packaging/pypi/pyproject.toml:45]: you should revert these changes

File: src/client/packaging/pypi/.bumpversion.cfg [hunk 3]
```diff
@@ -1,8 +0,0 @@
-[bumpversion]
```
Comment (melange396) [inline src/client/packaging/pypi/.bumpversion.cfg:1]: you should revert these changes (un-delete the file)

File: .github/workflows/create-release.yml [hunk 4]
```diff
@@ -27,10 +27,23 @@ jobs:
...
+            echo "\nInvalid version name: ${{ github.event.inputs.versionName }}"
+            exit 1
+          fi
+
+          if [[ ${{ github.event.inputs.versionName }} =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
+            NEXT_TAG="${{ github.event.inputs.versionName }}"
+          else
+            NEXT_TAG=$(bump2version --dry-run --list ${{ github.event.inputs.versionName }} | grep ^new_version | sed -r s,"^.*=",,)
+          fi
+          echo "next_tag=$NEXT_TAG" >> $GITHUB_OUTPUT
```
Comment (melange396) [inline .github/workflows/create-release.yml:46]: we need to make sure the new version gets applied:
```suggestion
          if [[ ${{ github.event.inputs.versionName }} =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then
            # use given version number
            NEXT_TAG="${{ github.event.inputs.versionName }}"
          else
            # calculate new version number based on given tag
            NEXT_TAG=$(bump2version --dry-run --list ${{ github.event.inputs.versionName }} | grep ^new_version | sed -r s,"^.*=",,)
          fi
          # apply given or calculated version number
          bump2version --new-version $NEXT_TAG _ignored_arg_
          # save version number for later
          echo "next_tag=$NEXT_TAG" >> $GITHUB_OUTPUT
```

**Review by melange396 — CHANGES_REQUESTED** [review]

File: .github/workflows/create-release.yml (hunk 4 above)
Comment (dshemetov) [reply .github/workflows/create-release.yml:46]: We never actually applied the version in this step before, so I don't think is right. The actual version bump happens in the  "Bump Python versions if client files changed" step below.

Comment (melange396) [reply .github/workflows/create-release.yml:46]: we did, it was in old line 33: `bump2version --list ${{ github.event.inputs.versionName }} | grep new_version | sed -r s,"^.*=",,`

the step below is just for the client version, this is for the server and acquisition code.

Comment (dshemetov) [reply .github/workflows/create-release.yml:46]: That line doesn't actually bump version, it just extracts the version number out and I'm already doing it in new line 44. The 
```
          # apply given or calculated version number
          bump2version --new-version $NEXT_TAG _ignored_arg_
```
in your suggestion was never done in the original action in this step and I don't think we need it. Are you intending to change the behavior of the original CI here? This PR is just about fixing the CI argument validation.

Comment (dshemetov) [reply .github/workflows/create-release.yml:46]: Ah I misunderstood, thought that `bump2version --list ${{ github.event.inputs.versionName }} | grep new_version | sed -r s,"^.*=",,` didn't actually operate on the files and make changes.

File: requirements.dev.txt [hunk 5]
```diff
@@ -1,5 +1,6 @@
 aiohttp==3.9.4
 black>=20.8b1
+bump2version==1.0.1
```
Comment (melange396) [inline requirements.dev.txt:3]: you can take this back out; it doesnt get used anywhere in this repo except for those two gh actions files, where it gets installed w/o versioning by calls to `python -m pip install bump2version`

Comment (dshemetov) [reply requirements.dev.txt:3]: cool, thanks for checking

**Comment by dshemetov** [comment]
Cleaned up the commit log by squashing the edits.
